### ISSUE TITLE
Highlight should not BLACK FLASH as default

### ISSUE THREAD

**Issue by offset-torque** [koreader/koreader#7854]
**INTRO**
I have a Kobo Libra H2O with an **E-ink Carta** screen. I set **FULL REFRESH RATE** to **NEVER** because screen stays clean between page turns. I even read through several chapters without needing a refresh. This screen is not an exceptional technology. E-ink Carta screens are 7 years old. Actually all current e-ink readers uses this type of screen. It is impossible to buy anything older than this generation.

I also enabled **AVOID MANDATORY BLACK FLASHES IN UI** and disabled **FLASH BUTTONS AND MENU ITEMS** and **FLASH KEYBOARD** so there should not be any black flash for me.

**PROBLEM** 
But highlighting action insists on black flashing everything I select (and unselect). It especially becomes awful and slow when highlighting a paragraph that cross page boundaries. And even worse if there is another highlight on the other page. It slowly flashes parts of screen that it feels like I have an ancient reader (I had one so I can compare).

Since I read academically, I use highlighting more. In the current book I am reading, I am at %25 and already have more than 60 bookmarks. And it is not exceptional to use e-ink readers this way since we now have bigger size e-readers especially for this purpose. So I guess other users will benefit from this change too.

Considering lithium ion batteries lose capacity after 2-3 years, I can make a guess that KOReader user base with old screen technology that needs black flashing (in other words devices older than 7 years) should be very small. I think black flashing should not be a default option while highlighting. Why unnecessarily restrain our new devices this way. If I remember correctly Plato reader highlighting behaves like I described and its selection feels much more fluid.

**SOLUTION**
Users should have at least one of these options IMHO:
- Highlighting action doesn't BLACK FLASH as default or
- Highlight menu can have an option to disable BLACK FLASH or 
- KOReader can disable BLACK FLASH automatically if user sets **FULL REFRESH RATE** to **NEVER**

(Images do black flashing too but since they are not interactive, it doesn't affect user experience that much)

**Comment by NiLuJe** [comment]
They actually don't.

It's a device-specific behavior (generally, only EPDCv1 REAGL Kindle can sometime help alleviate it, though) stemming from the technical limitations of displaying a gray shade.

**Comment by NiLuJe** [comment]
Simple test-case:

```
fbink -k width=200,height=50 -B WHITE
fbink -k width=200,height=50 -B GRAYA
fbink -k width=200,height=50 -B BLACK
```

Those will never request a flash, but you'll always see the capsules going black when switching WHITE <-> GRAY.

**Comment by NiLuJe** [comment]
The actual workaround would be to use a pure inversion for the CRe highlights (e.g., white on black), but I don't quite recall if that's doable right nwo (ping @poire-z).

**Comment by offset-torque** [comment]
> TL;DR: That's just how eInk works. 

Understood. That's sad.

> The actual workaround would be to use a pure inversion for the CRe highlights (e.g., white on black), but I don't quite recall if that's doable right now (ping @poire-z).

Yes I just checked Plato and it does like this. I was remembering partially correct. (It doesnt flash but stays black)

In highlights we have an option of **UNDERLINE**. It would be nice if highlighting just do underlining (while selecting I mean). That doesn't need black flashing whole line, right ? Is it possible just the underline can be reverse flashed ?

**Comment by poire-z** [comment]
> The actual workaround would be to use a pure inversion for the CRe highlights (e.g., white on black), but I don't quite recall if that's doable right now 

No, fillRect() and then draw the text - is hardcoded in crengine for text selection in progress. There is no "inverse text color", so choosing pure black for the filling color will have black on black.
It has a few more options to handle its own bookmarks (underline/solid), and a hardcoded `#ifdef CR_USE_INVERT_FOR_SELECTION_MARKS` (which is actually for bookmarks, and not selection marks).
Quick try at testing how switching them failed :) Not really interested with spending time on this.

Anyway, I'm quite sensitive to gray lines and gray backgrounds in our UI, flashing white>black>gray and gray>black>white - but well, I don't care that much about highlights.

Nothing much to do about current selection, but you could at least switch your other already-made (or only future?) highlights (drawn by frontend) to use underline, and see how this feels.

**Comment by NiLuJe** [comment]
@poire-z: Yeah, that's what I vaguely remembered, thanks for the pointers ;).

**Comment by offset-torque** [comment]
Considering we can turn pages without forced black flashing, I thought highlight reversing is an aesthetic decision. I didn't know we still have hardware layer restrictions. Maybe next versions of e-ink will be better.

**Thanks for answers NiLuJe and poire-z. Have a nice day.**
